model/interactive/src/main/proto/beam_interactive_api.proto

```diff
@@ -0,0 +1,178 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Protocol Buffers describing a service that can be used in conjunction with
+ * the TestStream class in order to control a pipeline remotely.
+ */
+
+syntax = "proto3";
+
+package org.apache.beam.model.interactive.v1;
+
+option go_package = "interactive_v1";
+option java_package = "org.apache.beam.model.interactive.v1";
+option java_outer_classname = "BeamInteractiveApi";
+
+import "beam_runner_api.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/duration.proto";
+
+
+service InteractiveService {
+  // A TestStream will request for events using this RPC.
+  rpc Events(EventsRequest) returns (stream EventsResponse) {}
+
+  // Starts the stream of events to the EventsRequest. It is also allowed for
+  // setting the playback_speed while RUNNING. If the stream is paused, this
+  // will unpause the stream at the specified playback speed.
+  rpc Start (StartRequest) returns (StartResponse) {}
+
+  // Advances the stream to the specified offset then pauses the stream. This
+  // starts the stream if it is not RUNNING then pauses the stream when the
+  // offset is reached.
+  rpc Advance (AdvanceRequest) returns (AdvanceResponse) {}
+
+  // Stops and resets the stream to the beginning.
+  rpc Stop (StopRequest) returns (StopResponse) {}
+
+  // Pauses the stream of events to the EventsRequest. If there is already an
+  // outstanding EventsRequest streaming events, then the stream will pause
+  // after the EventsResponse is completed.
+  // To un-pause, send either a Start or Advance request.
+  rpc Pause (PauseRequest) returns (PauseResponse) {}
+
+  // Sends a single element to the EventsRequest then pauses the stream.
+  rpc Step (StepRequest) returns (StepResponse) {}
+
+  // Responds with debugging and other cache-specific metadata.
+  rpc Status (StatusRequest) returns (StatusResponse) {}
+}
+
+// The state of the InteractiveService. The default state is STOPPED.
+enum State {
+  // The InteractiveService is not replaying. Goes to RUNNING with a
+  // StartRequest.
+  STOPPED = 0;
+
+  // The InteractiveService is replaying events. Goes to PAUSED with a
+  // PauseRequest. Goes to STOPPED with a StopRequest.
+  RUNNING = 1;
+
+  // The InteractiveService is paused from replaying events. Goes to RUNNING
+  // with either a StartRequest or a StepRequest. Goes to STOPPED with a
```

Comment:
Does it go to STEPPING (instead of RUNNING) with a StepRequest?
Also, does it go to ADVANCING with an AdvanceRequest?

Reply:
Gotcha, I rewrote the state comments to be very explicit about the state machine and what happens at each state.